src/libraries/Microsoft.Extensions.Configuration/tests/FunctionalTests/Microsoft.Extensions.Configuration.Functional.Tests.csproj

```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>$(NetCoreAppCurrent);net461</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent);net461-windows</TargetFrameworks>
```

Comment:
Can you explain the logic behind the changes on netfx configurations and why we are adding OS to some of them? Is the one here due to the package reference to a windows-only package?

Reply:
Libraries which projectreference rid-specific libraries need to be rid-specific as well. In this case, all the libraries that I annotated incorrectly referenced the `netstandard2.0` asset instead of the `net461-windows` asset. The technical reason behind that is that `net461` can't reference `net461-windows`, so `netstandard2.0` wins.

I discovered that when bringing back System.Security.AccessControl and had to make changes to use it as a live reference.

Reply:
I see, this seems like a very fragile system when trying to get them correct. Should we instead do one of the following:
- Transform all net461 configs to instead be net461-windows, that way all configurations are referencing the right live libs.
- Add some build checks that check P2Ps when building netfx libs, and ensuring that we are not incorrectly getting a NS2.0 asset when the P2P also has a net461-windows asset.

Of course we don't have to fix it here, but I thought of bringing the suggestions here anyway

Reply:
> I see, this seems like a very fragile system when trying to get them correct. Should we instead do one of the following:

Agree with this being a fragile design. It predates me, it existed for several years and apparently it worked just fine. Even though this is hard to get right without understanding RID compatibility rules, it's the closest / best implementation based on the rid graph.

> Transform all net461 configs to instead be net461-windows, that way all configurations are referencing the right live libs.

Reason why we didn't do this yet is because of these two major downsides:
- Making a tfm rid specific without also building a rid agnostic tfm of the same TargetFrameworkIdentifier results in  packages having duplicated assets in the i.e. lib/net461/ and runtimes/win/lib/net461/ folders. For more information see https://github.com/dotnet/runtime/blob/c94b75f662030a09c8747d79a16d9f4f2c9d44f1/eng/packaging.targets#L104-L108.
- net461-windows is a tfm alias that nuget doesn't understand and the `-windows` part needs to be stripped from the target framework alias: https://github.com/dotnet/arcade/blob/761d2e6545a31d2ebf9cb60443ed2a9c2f8268b5/src/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk/src/build/Microsoft.DotNet.Build.Tasks.TargetFramework.Sdk.props#L21-L22. Adding the `-windows` rid to every .NETFramework tfm results in this hack being applied in many more places and more NuGet.config files being necessary to disable VS's restore which can't handle restoring them: https://github.com/dotnet/runtime/blob/c94b75f662030a09c8747d79a16d9f4f2c9d44f1/src/libraries/System.Data.OleDb/NuGet.config#L2.

> Add some build checks that check P2Ps when building netfx libs, and ensuring that we are not incorrectly getting a NS2.0 asset when the P2P also has a net461-windows asset.

Correct me if I'm wrong but I believe the [ProjectReference protocol](https://github.com/dotnet/msbuild/blob/main/documentation/ProjectReference-Protocol.md) doesn't expose the necessary information to implement such a check. Of course you could always implement custom targets and MSBuild calls but that won't work with static graph build which we want to leverage.

Reply:
> I see, this seems like a very fragile system when trying to get them correct. 

I believe at one point package validation would catch any mistake here and would actually drive folks to do the right thing / teach them about the RID specific asset being preferred. Not sure if that's still the case.

Reply:
> I believe at one point package validation would catch any mistake here

@ericstj can you please elaborate on "at one point ... would catch any mistakes here"? Do you mean that the new PackageValidation infrastructure will eventually catch this in the future, did you refer the the old pkgproj system or to the runtime repo specific testPackages infra (which creates wrapper projects and restores the package)?

Reply:
For the sake of this discussion, I filed https://github.com/dotnet/runtime/issues/58495 to simplify .NETFramework tfms.

Reply:
I'm talking about package testing that we run that consumes the nuget packages we build in a project.  We would run that with RIDs on .NETFramework, so it would see duplicate runtime types in the case someone forgot to include a RID specific asset, and would see duplicate reference types if someone forgot the RID-less asset.  I *think* we may still catch this today if we use the runtimeTargets on .NETFramework.

Previously we would also sometimes catch this in package build itself if the assembly versions differed (as is the case in servicing).

I added a comment to the issue about making sure we have test coverage of this scenario.